gpbft/gpbft.go

```diff
@@ -918,6 +948,66 @@ func (i *instance) broadcast(round uint64, step Phase, value ECChain, createTick
 	}
 
 	_ = i.participant.host.RequestBroadcast(mb)
+	i.broadcasted.record(mb)
+}
+
+// tryRebroadcast checks whether the conditions for rebroadcasting messages are
+// met, and if so rebroadcasts messages from current and previous rounds.
+// Rebroadcast is needed when both of the following conditions are met:
+//   - There has been no progress since latest broadcast.
+//   - Rebroadcast backoff duration has elapsed.
+//
+// If there has been progress since latest broadcast this function resets the
+// number of rebroadcast attempts and returns immediately.
+func (i *instance) tryRebroadcast() error {
+	if i.progressedSinceLatestBroadcast() {
+		i.rebroadcastAttempts = 0
+		return nil
+	}
+
+	if i.rebroadcastAttempts == 0 {
+		// It is the first time that rebroadcast is necessary; set initial rebroadcast timeout.
+		i.rebroadcastTimeout = i.phaseTimeout.Add(i.participant.rebroadcastAfter(0))
+	}
+
+	if i.rebroadcastTimeoutElapsed() {
+		err := i.rebroadcast()
+		i.rebroadcastAttempts++
+		i.rebroadcastTimeout = i.rebroadcastTimeout.Add(i.participant.rebroadcastAfter(i.rebroadcastAttempts))
+		return err
+	}
+
+	// Schedule the next rebroadcast.
+	i.participant.host.SetAlarm(i.rebroadcastTimeout)
```

Comment:
I spent some time thinking there was a problem with overriding the alarm set for a phase timeout here. I think I've realised that this assumes that the phase timeout has already passed, and that the alarm has been delivered. I think the phase timeout having passed is true in the call paths, but are we sure that the alarm has been delivered (or if not, that it wouldn't do anything)?

Working through out loud:
1. The participant begins the prepare phase, sends a message, and sets an alarm for phase timeout.
2. The participant receives a prepared message from someone else. The phase timeout has not elapsed, so this method should not be called.
3. The participant receives a message right on the phase timeout. The message gets to the GPBFT code before the alarm (maybe a race in the host). In tryPrepare, the participant notices that the phase has timed out, so invokes this method. There is no progress since last broadcast, and rebroadcast attempts is zero, so the rebroadcast timeout variable is set. An alarm is then set for that rebroadcast timeout, which overrides the phase timeout alarm that would have been delivered immediately next.

This seems technically wrong, but inconsequential. If the alarm had arrived first, the participant still would have ended up here and set a new rebroadcast timeout. It didn't miss any opportunity to make progress based on the phase timeout.

Reply:
Good point thank you

Reply:
I have resolved this by only setting the rebroadcast alarm only when it is unset.